### DISCUSSION TITLE
Including pages from a PDF

### DISCUSSION THREAD

**Discussion by simolus3** [sile-typesetter/sile#1515] [Q&A] [answered]
I'm looking for a way to include pages from an existing PDF into a document created by sile, similar to `\includepdf` from the pdfpages latex package.
I'm aware that I can insert a PDF as an image, but I'm looking to insert (potentially multi-page) PDFs as into a document without modifications or any sile frames being typeset on those pages.

Currently I'm doing post-processing with `pdfunite` after I ran `sile`, but I'm wondering if there's a way to simplify this process by doing it all in one step. So, is there a way to do such thing in sile that I'm not seeing?

**Comment by alerque** [comment, marked as answer]
I don't think you've missed anything obvious, I don't think we have an easy mechanism for this. I haven't looked yet, but *if* the libtexpdf library has a mechanism for writing a page from a blob (basically a passthrough) it might not be too hard to implement this. It would require a fiddling with the outputter a little, but potentially not out of reach of a package with a few lines of lua. If the libtex library has this functionality but we haven't surfaced the API in our Lua wrapper (my best guess says this is going to be the case) then it would require a small change to the C code to surface the feature, then again as above a small Lua package to make it accessible to a document.

In my own production work I currently do something similar to you — except I use `pdftk` as my post-processing weapon of choice instead of `pdfunite`. It allows a bit more control over page ranges so I can splice together multiple pages in the middle of other documents or pages from one document to the front and back of another (e.g. for cover pages).

**Reply by simolus3** [reply]
Thanks for the suggestions! It looks like an acceptable way to do this may be to repeatedly call this with different page numbers here, the function seems to support extracting specific pages as images from a PDF file.

https://github.com/sile-typesetter/sile/blob/3f80a9bfd4831cc7856cf2c9eb5c63f1caae432b/src/justenoughlibtexpdf.c#L344

I'm not too familiar with the internals but I'll try exposing that in a command and see how much breaks.

**Reply by Omikhleia** [reply]
@simolus3 One needs to take into account the image size method. Unless you really want to give it a try, I am pretty sure I could try and add this to #1346 -- It touches the same bits of code too.

**Reply by Omikhleia** [reply]
N.B. Yes it works:
![image](https://user-images.githubusercontent.com/18075640/183141095-c77afa88-3bcc-4677-af1c-356f32749022.png)

(SILE logo in PDF just to check I didn't broke it, and then 3 pages from another document)

**Reply by alerque** [reply]
Maybe I misunderstood what was being asked here, I thought the goal was to output them *as* pages not embed the pages as nested objects in the page SILE is working on. No?

**Reply by Omikhleia** [reply]
Ah indeed perhaps! So to include all pages (or a range) of a document in another.
I read too quickly.

Still, being able to select a page on a PDF `\img` might be interesting too.

**Reply by alerque** [reply]
Yes, selecting a page from `\img` would be nice too ;-)

**Reply by simolus3** [reply]
> Maybe I misunderstood what was being asked here, I thought the goal was to output them as pages not embed the pages as nested objects in the page SILE is working on. No?

Yes, that is what I'm trying to do :) My idea was essentially to get the page count of a PDF file via libtexpdf, then push a new page for each page in the pdf I want to include. For each page, I'd call `texpdf_doc_set_mediabox` to get the page size to match and then include the page as a full-page image. I'm fine with references in the included pages not working, I kind of hope that text can still stay selectable with this method but I'm not sure.

**Reply by alerque** [reply]
References would be a major challenge, we'd have to invent a PDF *reader* and try to integrate that stuff, but PDF's don't neatly contain stuff like references in "pages". Lets leave that out of the discussion for now.

The page size matching is not what I was assuming this would entail. Rewriting the media box would also involve some level of post processing. I was assuming it would just be added inline and then have SILE virtually skip over the page numbers.

But that does ping up the question: why now just come up with a function that uses a new master frame layout with no frames (or rather only the page frame), starts a new page, switches to that master, outputs an `\img[width=100%pw,height=100%ph]`, then starts a new page and goes back to the previous frame layout?

**Reply by davidchisnall** [reply]
Is there an easy way to do this now?  It's convenient for PDF eBooks to be able to prepare the cover image in another program and then insert it as the title page.  It can also be useful for things like copyright pages, where they often want specific layouts and doing them in something other than the typesetting package can be useful.  Doing this after the initial PDF has been generated breaks the table of contents and other PDF metadata.  There's a nice pdflatex package for doing this that I've used in the past but I'm struggling to replicate the same thing in SILE.

**Reply by Omikhleia** [reply]
@davidchisnall inserting a given page from a PDF now works,[^1] with `\img[src=mydef.pdf, page=1]` (or whatever page number). Of course you'll have to set the page frame appropriately and to use width/height on the image for appropriate scaling.

I also added image support on the `\background` command (from the same named package), which would be another option (probably preferable?) - but I don't think I added page selection here (so only the first page of a PDF can be picked. I didn't test with a PDF either, though I'd expect it to work too here)

[^1]: With a small caveat, some versions of PDF aren't supported by the backend (libtexpdf), see #1845

**Reply by Omikhleia** [reply]
(As for inserting multiple pages from a PDF - Well no, this has not been done yet.)

**Reply by alerque** [reply]
Doing a multi-page version would be about 5 lines of Lua (untested just off the cuff):

```lua
SILE.registerCommand("manypages", function (options, content)
   for i = options.start, options.end do
      SILE.call("img", { options.src, page = i })
   end
end
```

```sile
\manypages[src=myfile.pdf,start=2,end=6]
```

**Reply by davidchisnall** [reply]
Thanks.  I used \background then \par\eject to get a title page.  With resilient, I also needed \nofolios to prevent it from putting a page number on the title page (I'm using the resilient document class but not all of their extra complexity).

**Reply by Omikhleia** [reply]
You might also need to reset the page counter after your page break, if you do not want the title/cover page to be counted (e.g. `\set-counter[id=folio,value=0`). That, and the `\nofolios`, are not specific to resilient by the way, the same would apply with the core "book" class. (Not sure what the "extra complexity" is, but that's another topic.)

**Reply by davidchisnall** [reply]
> You might also need to reset the page counter after your page break, if you do not want the title/cover page to be counted

I actually do want it to count, because I find it annoying when the PDF page number and the page numbers in the text are different.  For the print flow, I won't include the cover since it gets added differently, which means the page numbers in the PDF and print editions won't quite line up, but the ePub and web versions don't have page numbers at all, so that's probably fine.

> That, and the \nofolios, are not specific to resilient by the way, the same would apply with the core "book" class

Ah, that explains why I couldn't find them in the resilient docs.  I was confused because there was a \noheaders command but no \nofooters command, but looking at the code for the resilient class I saw it used \nofolios in the start and tried that.  

> Not sure what the "extra complexity" is, but that's another topic

You have an opinionated build system that is probably very useful for people who want to target only resilient but is just inconvenient for a flow that targets multiple back ends, of which SILE is one.  Being able to output a single XML file that I feed to SILE for it to consume and produce a PDF is great.

For reference, [here is the output of the HTML flow](https://cheriot.org/book/), happy to share the SILE-generated PDF with anyone who is interested.  It's now in quite a nice state, I will need to either add some float support (pointers to where in the code I should look welcome!) or manually move a load of floats after I get it back from the copyeditor.  Aside from that the index is pretty terrible, but that's a content problem not a typesetting one.

**Reply by Omikhleia** [reply]
> You have an opinionated build system that is probably very useful for people who want to target only resilient but is just inconvenient for a flow that targets multiple back ends, of which SILE is one. 

I'd probably disagree (both in scope and purpose) but this is not the right place anyway ;)
